LanguageFeatures/Extension-types/exhaustiveness_object_A01_t02.dart

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Switch statements and expressions with a sealed class as a
+/// matched type are always exhaustive
+///
+/// @description Check that it is no compile-time error if the matched value
+/// type of a switch statement is an extension type with a sealed class as a
+/// representation type and the set of cases is exhaustive
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inline-class
+
+import "../../Utils/expect.dart";
+import "../Patterns/Exhaustiveness/exhaustiveness_lib.dart";
+
+extension type FaceET1<T>(Face<T> _) {}
+extension type FaceET2<T>(Face<T> _) implements Face<T> {}
+
+String test1_1(FaceET1 face) {
+  switch (face) {
+    case Jack(): return 'Jack';
+    case Queen(): return 'Queen';
+    case King(suit: _): return 'King';
+  }
```

Comment:
This one is blocked on https://github.com/dart-lang/language/issues/3534 and https://github.com/dart-lang/language/issues/3524: If we don't consider `FaceET1` and `FaceET2` as always-exhaustive then line 27 would be considered reachable, and we'd have an error. If we don't consider the extension type erasure of `FaceET1` or `FaceET2` then we can't conclude that the switch statement is exhaustive.

I believe this implies that we can only expect this to compile with no errors if both of those proposals are accepted.

Reply:
@eernstg I believe that `FaceET1` and `FaceET2` are considered as always exhaustive now. So, this shouldn't be blocked any longer. But why we should have an error here? `Face` is a sealed class and `Jack`, `Queen` and `King` is an exhaustive set of descendants. The current implementation doesn't produce an error here

Reply:
Ah, that's probably because flow analysis trusts a switch statement to be exhaustive when it has been handled by the exhaustivenses algorithm. So with 3524, `FaceET1/2` are always-exhaustive, the exhaustiveness check is satisfied, and the flow analysis doesn't care that it can't see the same thing, it just concludes that we can't reach the end of the function because every branch of this switch will complete returning (that is, it will not complete normally).

So we're good!